### DISCUSSION TITLE
New - Document List Component

### DISCUSSION THREAD

**Discussion by owensgit** [Legal-and-General/canopy#744] [Canopy requests]
*Type of request*
New component / pattern

*Component, pattern or foundation?*
Component or Pattern - not sure which :) 

*Nature of request*

This was originally intended to be added to Canopy, but due to time pressures it was instead implemented as a custom component.

I'd like to now add this to Canopy as a new component.

It renders a list of documents that the user has, with the ability to show unread documents.

Screenshot shows the current custom build.

<img width="972" alt="Screenshot 2022-03-18 at 15 58 07" src="https://user-images.githubusercontent.com/1943532/159038227-aa12188e-362e-411e-a8b0-48be674c73d4.png">

*Proposed API*

```
<lg-document-list>
  <lg-document-list-item
    [read]="true"
    [url]="documentUrl"
  >
    <lg-icon name="pdf"></lg-icon>
    {{documentName}}
    <lg-document-list-item-date [date]="unixTimestamp">
      <!-- can also manually add text if needed, 
           but date input preferred -->
    </lg-document-list-item-date>
    <lg-document-list-item-size>
      100 - 500kb
    </lg-document-list-item-size>
  </lg-document-list-item>
</lg-document-list>
```

**Comment by elenagarrone** [comment]
Thanks for submitting the request @owensgit 👍 

My first thought is: do we need a specific component for this? why not use a simple Canopy table and create a pattern? That's how it is in one of our project:
![Screenshot 2022-03-22 at 09 09 09](https://user-images.githubusercontent.com/8397116/159446111-afcb7e51-2c94-4142-ba8a-b8745f517e39.png)

If we go with your approach how would you develop it semantically?
My assumption is that it would be:

- `lg-document-list` => `role="list"`
- `lg-document-list-item` => `role="listitem"` and have an inner button that wraps the whole content?

And the components would have to add some access text e.g. the `lg-document-list-item` should add text like `PDF` before or after the document name, the `lg-document-list-item-size` should say `Size: <size-data>`, etc.

Also, on another note I don't think the dot for "read/unread" is very intuitive 🤔

**Reply by alexcanning** [reply]
To add to Elena's commentary on the design, the original design was supposed to be a quick fix. We didn't give it much thought from a Canopy point of view as Canopy was still in its infancy. I have suggested to @michelleallendavey that either:
1. We simply use a table (like Elena's example) which would have better accessibility and already exists
2. We go back to the drawing board and push the design a bit further, to distinguish it from the table

**Reply by michelleallendavey** [reply]
Yeah – though a custom design might look nicer and have more scope for custom interactions etc, it feels like it makes sense from an accessibility and consistency perspective to have it as a table... though I'm not particularly precious either way. Do you know if there was a reason it wasn't implemented as one to begin with @owensgit?

**Reply by owensgit** [reply]
@elenagarrone thanks for the feedback, I agree with the accessibility solutions 👍 

@michelleallendavey I'm not sure why it was implemented like this originally, but I suspect it was just down to just looking nicer than a table as you suggested. In terms of accessibility, a custom design can be made accessible using solutions as suggested above, so we can definitely do that if you guys prefer that kind of design.

**Reply by michelleallendavey** [reply]
Cool, thanks @owensgit! 

@alexcanning and I just had a quick chat about it – I think we've got so many Canopy things going on at the moment that we don't have time to give a custom solution the time it deserves right now... so perhaps a table is the way to go for now? Appreciate that might feel like a step backwards in Annuities though. Essentially the dilemma for me is – no time to create designs for custom component, but will need to implement a doc table in my current product team in the next couple of sprints 😅

**Reply by owensgit** [reply]
If you're doing it anyway as part of upcoming work, maybe best to wait for that, I can then use that new component for in our existing product when ready 👍 

Or if you don't have time for that in your team, then the team could use the existing doc list component already in the codebase (as currently used) or go with the table component. 

Another idea could be to implement this new component in Canopy as above (I'd be happy to do that), use that everywhere for now, and then come back to it for an design refresh later.

Let me know what you think

**Reply by michelleallendavey** [reply]
Yeah I think we're going to have limited time in my current product team for this so potentially either using the existing non-Canopy component, or making this a Canopy component that we come back to later would be a good option for us I think... (though I suspect Alex might have something to say about the latter plan 😃

If we were to use a table, are we able to make a whole row clickable like it is in the existing non-Canopy component? Or would it have to be a link like the example Elena added above?

**Reply by owensgit** [reply]
Thanks @michelleallendavey, I'm happy to convert the existing non-Canopy component into Canopy so it can be used by both teams, and others. It seems like the right thing to do considering it's already needed for two produc types and there won't be any time seem to revisit the design. @alexcanning @elenagarrone are you also happy with this approach?

> If we were to use a table, are we able to make a whole row clickable like it is in the existing non-Canopy component? 

It's possible, but I don't think that's a good idea for a table, given that users may want to copy and past data from a column - e.g. payment history table containing dates, amounts, descriptions etc. If we were to use a table, I would recommend just making the title clickable.

**Reply by elenagarrone** [reply]
@owensgit form a development point of view, if you have time to consider accessibility for it then I think we can add it.

Instead of the list role we could use the table role: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/table_role
Also from an accessibility point of view we should add the `read/unread` as a new column.

I don't think just copy pasting the non-canopy component is enough. We should make it better if we add it to the library :/ otherwise it's easier to use the table which is more accessible.

**Reply by alexcanning** [reply]
My preference would be to not convert the existing non-Canopy component for the reasons articulated previously - it was a placeholder which is not distinct enough from the table component. If we're not designing something wholly different, we should just use a table with traditional links for document titles.

**Reply by owensgit** [reply]
Cool thanks all, I'll leave the originally component as-is for now 👍

**Comment by owensgit** [comment]
**Decision: Do no implement**
This component was orignally a placeholder and not intended for wider use. The design team would like to revisit it to fully consider the final design and accessibility. For now teams should use the existing Table component.
